### PR TITLE
added security

### PR DESCRIPTION
**Pull request by fmIst0**
Added User entity,
Added Role entity,
Added the required DTOs and mapper,
Added AuthenticationController with the following methods:
- public UserLoginResponseDto login(UserLoginRequestDto request)
- public UserResponseDto register(UserRegistrationRequestDto request) throws RegistrationException,
Added FieldMatch annotation for checking that password and repeatPassword field in the UserRegistrationRequestDto are the same
RegistrationException and EntityNotFoundException handled using the GlobalExceptionHandler,
Addedthe JWT support to your application

### CODE REVIEW

File: src/main/java/com/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+public class UserRegistrationRequestDto {
+    @Email(message = "Email is not valid",
+            regexp = "^[a-zA-Z0-9_!#$%&'*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$")
+    @NotEmpty
+    private String email;
+    @NotBlank
+    @Size(min = 6, max = 100)
+    private String password;
+    @NotBlank
+    @Size(min = 6, max = 100)
```
Comment (Rommelua) [inline src/main/java/com/bookstore/dto/user/UserRegistrationRequestDto.java]:
```suggestion
```

File: src/main/java/com/bookstore/dto/user/UserLoginRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+public class UserLoginRequestDto {
+    @Email(message = "Email is not valid",
+            regexp = "^[a-zA-Z0-9_!#$%&'*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$")
```
Comment (Rommelua) [inline src/main/java/com/bookstore/dto/user/UserLoginRequestDto.java]: do we need this regexp?

File: src/main/java/com/bookstore/model/Role.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Getter
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE roles SET is_deleted=true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "roles")
+public class Role {
```
Comment (Rommelua) [inline src/main/java/com/bookstore/model/Role.java:24]: add equals/hashCode

File: src/main/java/com/bookstore/repository/user/UserRepository.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import com.bookstore.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+    @Query("FROM User u JOIN FETCH u.roles r "
+            + "WHERE u.email = :email "
+            + "AND u.isDeleted=FALSE "
+            + "AND r.isDeleted=FALSE")
```
Comment (Rommelua) [inline src/main/java/com/bookstore/repository/user/UserRepository.java]:
```suggestion
    @EntityGraph(attributePaths = "roles")
```

File: src/main/java/com/bookstore/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        user.setShippingAddress(requestDto.getShippingAddress());
+        Role userRole = roleRepository.findRoleByName(Role.RoleName.USER)
+                .orElseThrow(() -> new RegistrationException("Can't find role by name"));
+        Role adminRole = roleRepository.findRoleByName(Role.RoleName.ADMIN)
+                .orElseThrow(() -> new RegistrationException("Can't find role by name"));
+        Set<Role> roles = new HashSet<>();
+        if (user.getEmail().startsWith("admin")) {
+            roles.add(userRole);
+            roles.add(adminRole);
+        }
```
Comment (Rommelua) [inline src/main/java/com/bookstore/service/impl/UserServiceImpl.java]:
```suggestion
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by aaananas — APPROVED** [review]

**Pull request merged**